Change link underline technique to linear-gradient

Scotchester opened the pull request:
If I did this right, @himedlooff, all you should need to do is pull this down, re-run `npm install` and `grunt vendor`, and rebuild the docs.

Let me know how it goes.

Scotchester committed:
Add explicit :link styling on link color mixin

Scotchester committed:
Comment out existing link border rules

Scotchester committed:
Change underline technique to linear-gradient

himedlooff commented on src/cf-base.less:
```diff
@@ -315,39 +315,57 @@ figure {
     }
 }
 
+// To be documented: mixin for simulating underlines that break around descenders
```
You've got a "To be documented" note here. Can you add that documentation please?

Scotchester replied:
Converted to `.u-` mixin and moved to `cf-utilities.less`.

himedlooff commented on src/cf-base.less:
```diff
@@ -417,16 +435,24 @@ a {
+    text-decoration: none;
+
+    .cssgradients & {
+        background-size: 0 0;
```
No need for double zero, is there?

Scotchester replied:
Good catch.

himedlooff commented on src/cf-base.less:
```diff
@@ -417,16 +435,24 @@ a {
+        text-decoration: underline;
+
+        .cssgradients & {
+            background-size: (2px + 2px) 2px;
```
Is there somewhere that explains the significance of `(2px + 2px)`? I'm assuming one is for the width of the underline and maybe the other is an offset from the text?

Scotchester replied:
Yeah, dot width + spacing. The sum needs to match the sum of the values in the `linear-gradient` declaration. I've converted all of these to variables.

himedlooff commented on src/cf-base.less:
```diff
@@ -315,39 +315,57 @@ figure {
         color: @pacific-50;
+        
+        .cssgradients & {
+            background-image: linear-gradient(to right, fade(@pacific-50, 50%) 2px, transparent 2px);
```
So changing the color of the underline is intentional right?

Scotchester replied:
Yeah. I thought with the extra dot size, the color could be a little more subtle. I will discuss all of my proposed design changes with the designers at the DM review on Wednesday.

Scotchester committed:
Updates from code review

himedlooff commented:
We should offer a way to break the underlines without having to use a mixin. Like maybe `.u-breaking-underlines` as a utility class.

Also the demo doesn't seem to be working right. The first link looks ok but the rest don't appear to have the breaking underline, or it's very hard to tell.

The "Breaking link underlines mixin" demo

![image](https://cloud.githubusercontent.com/assets/1093373/5729160/163b761e-9b3e-11e4-9e0f-b0fbedeaa46d.png)

![image](https://cloud.githubusercontent.com/assets/1093373/5729152/026d3b18-9b3e-11e4-8d92-00d6a00ea8e5.png)

Scotchester commented:
I considered doing that. The prospect of changing background colors worries me slightly, but it probably makes sense to have a utility class with a default of white. Perhaps including white in the name would be good for awareness.

Hmm, yeah. The breaking effect is not as good as I thought, either. It's working, just not well enough for some letters. Maybe a difference in browser rendering. What are these screenshots from?

himedlooff commented:
Chrome

Scotchester commented:
Let's roll with it as-is for now, since it's not the core purpose of this change. In the near future, I'll investigate possible optimizations to the breaking technique, starting with this excellent article: https://eager.io/blog/smarter-link-underlines/

himedlooff commented:
I'd like for the breaking underlines to be more than a mixin though. That should be easy to add right?

Scotchester commented:
Sure, I'll add that. `.u-break-underlines-on-white`?

himedlooff commented:
I think `.u-break-underlines` is sufficient.

Scotchester committed:
Update class name

Scotchester commented:
Updated.

himedlooff commented:
Thanks. Let's merge after the Design Manual meet up tomorrow.

himedlooff commented:
Status: waiting for others to play around with the demo as stated here: https://github.com/cfpb/design-manual/issues/244

Scotchester committed:
Improve breaking underlines, selected links

- Breaking underlines now interact better with serifed descenders
- Text shadow removed when selecting text that includes links
- Link color darkened by 10% when selecting text that includes links

anselmbradford commented:
Ping!

Scotchester commented:
Sorry, this is a really far-reaching change. I think it's on hold until we get all the DM surge work out the door. Then we can buckle down and address all of the places that this change will affect.

jimmynotjim commented:
Should we close it and start over? Things have changed quite a bit in this time

Scotchester commented:
Not so much has changed in cf-core that this PR isn't still useful. I still have an `underlines` branch locally that I'm working with.
